### PR TITLE
Document ADR 0008 about unrecognized fields

### LINKED ISSUE #1266: Should we accept metadata that includes unrecognised fields?
**Description of issue or feature request**:

The current reference implementation will ignore unrecognised fields in a metadata file when loading it. Furthermore, some TAPs (i.e. snapshot Merkle Tree) are relying on this behaviour to introduce new fields to the specification without making the metadata invalid for older clients which don't recognise the field.

**Current behavior**:

The current client implementation ignores unrecognised/unexpected fields.

The new metadata API currently does not support unrecognised fields. If a metadata file including an unexpected field is loaded, the load fails.

**Expected behavior**: 

To Be Determined.

We should decide (and capture in an ADR) whether we expect to be able to load metadata which includes additional fields beyond wha the implementation expects.

See also a related issue in the in-toto reference implementation: https://github.com/in-toto/in-toto/issues/186

### PR DESCRIPTION
**Pull request by MVrachev**
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

Fixes #1266 

**Description of the changes being introduced by the pull request**:
Even though, this ADR documents something already implied in the TUF
spec in [document formats](https://theupdateframework.github.io/specification/latest/#document-formats)
it seems better to document this decision clearly so that it could be
referenced and give an explanation why someone can load a metadata file
with additional unrecognized fields.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Review by mnm678 — APPROVED** [review]
This looks great, thanks @MVrachev! This should help us move forward with some of the TAPs

File: docs/adr/0008-accept-unrecognised-fields.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+Technical Story: https://github.com/theupdateframework/tuf/issues/1266
+
+## Context and Problem Statement
+
+The current reference implementation will ignore unrecognized fields in a
+metadata file when loading it.
+This leads to the side effect that if you read a metadata file with unrecognized
+fields and immediately write it back to the disk, this file will be modified.
+
+Furthermore, some TAPs ([6](https://github.com/theupdateframework/taps/blob/master/tap6.md), [10](https://github.com/theupdateframework/taps/blob/master/tap10.md), [14](https://github.com/theupdateframework/taps/blob/master/tap14.md), [15](https://github.com/theupdateframework/taps/blob/master/tap15.md), and [16](https://github.com/theupdateframework/taps/blob/master/tap16.md)) are relying on that unrecognized
```
Comment (joshuagl) [inline docs/adr/0008-accept-unrecognised-fields.md]: Nit: this is a long line, can we wrap it?

File: docs/adr/0008-accept-unrecognised-fields.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+fields will be accepted to introduce new fields to the specification without
+making the metadata invalid for older clients which don't recognize the field.
+
+## Decision Drivers
+- The TUF specification implies support for custom attribute-value fields
+See [Document formats](https://theupdateframework.github.io/specification/latest/#document-formats)
```
Comment (joshuagl) [inline docs/adr/0008-accept-unrecognised-fields.md]:
```suggestion
- The TUF specification implies support for custom attribute-value fields, see [Document formats](https://theupdateframework.github.io/specification/latest/#document-formats)
```

File: docs/adr/0008-accept-unrecognised-fields.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+- Backward compatibility
+- If we perform the following operations on a metadata file with no
+intermediate operations:
+1. read the metadata file
+2. write the metadata file back to the disk
+the checksum (the content) of the file must not be changed.
```
Comment (joshuagl) [inline docs/adr/0008-accept-unrecognised-fields.md]: this Markdown is rendered such that "the checksum (the content) of the file must not be changed" appears on the same line as "write the metadata file back to the disk"

**Review by joshuagl — APPROVED** [review]
Thanks for capturing this @MVrachev. I made a few minor comments, otherwise this looks good.

**Comment by MVrachev** [comment]
@joshuagl I addressed your comments and amend the commit so it can be ready to be merged.

PS: I had to force-push a couple of times to fix some formatting issues.

File: docs/adr/0008-accept-unrecognised-fields.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+- [TAP 14](https://github.com/theupdateframework/taps/blob/master/tap14.md)
+- [TAP 15](https://github.com/theupdateframework/taps/blob/master/tap15.md)
+- [TAP 16](https://github.com/theupdateframework/taps/blob/master/tap16.md)
+
+are relying on that unrecognized fields will be accepted to introduce new fields
+to the specification without making the metadata invalid for older clients who
+don't recognize the field.
+
+## Decision Drivers
+- The TUF specification implies support for custom attribute-value fields,
```
Comment (joshuagl) [inline docs/adr/0008-accept-unrecognised-fields.md]:
```suggestion
- The TUF specification implies support for unrecognised attribute-value fields,
```

File: docs/adr/0008-accept-unrecognised-fields.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+see [Document formats](https://theupdateframework.github.io/specification/latest/#document-formats)
+- Backward compatibility
+- If we perform the following operations on a metadata file with no
+intermediate operations:
+1. read the metadata file
+2. write the metadata file back to the disk
+
+then, the checksum (the content) of the file must not be changed.
+- Flexibility to add new fields in the spec without adding breaking changes.
```
Comment (joshuagl) [inline docs/adr/0008-accept-unrecognised-fields.md:35]: I think this is and the second bullet (backwards compatibility) are the same? This is just more descriptive of what we mean by that term. Let's remove the 2nd bullet?

**Comment by MVrachev** [comment]
I am not sure why, but we have a CI failure again on Windows
https://github.com/theupdateframework/tuf/pull/1343/checks?check_run_id=2342301986
@jku does this look familiar to other Windows CI failures?

**Comment by jku** [comment]
Not at all, this is tox calling pip and pip failing to download "typed-ast" in some fantastic way...

**Comment by jku** [comment]
For reference: the CI failure seems like a pypi problem (made worse by our decision to use pre-release build tooling so downloading more often). Disappeared on re-run, I'm not filing any issues on that. Merging this.

**Pull request merged**